dao/src/main/java/com/epam/esm/dao/TagDao.java

```diff
@@ -15,21 +14,31 @@
  */
 public interface TagDao {
 
+    /**
+     * Find Tag the most widely used tags of user with given id.
+     *
+     * @return list of Tag entities.
+     * @since 2.0
+     */
+    List<Tag> findMaxWidelyUsed(long userId);
```

Comment:
Please create criteria for it and use Specification and method 'findBy'

Reply:
I changed this method to Criteria API, but i couldn't moved it to findBy method, because findBy() uses a QueryBuilder that coustructs a simple single criteria without subselect e.t.c. My native query for finding most widely used tags have a subselect which has a order by and limit statements. This statements is not supported by subSelect() instance of Criteria API. So I had to separate my navite sql to 2 criteria queries. QueryBuilder's method return a single criteria for specified parameter in request, but this case contains 2 queries(subselect and mainselect). That is why I didn't move this logic into QueryBuilder's methods and left it in TagDaoImpl.
Commit 4e95386.